### PR TITLE
GAP-2092: store salt in db, add timeout error page for logging in

### PR DESCRIPTION
**Pull request by ConorFayleTCO**

### CODE REVIEW

File: src/main/java/gov/cabinetofice/gapuserservice/exceptions/NonceExpiredException.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package gov.cabinetofice.gapuserservice.exceptions;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.I_AM_A_TEAPOT)
```
Comment (chris-forbes) [inline src/main/java/gov/cabinetofice/gapuserservice/exceptions/NonceExpiredException.java]: why are we returning this response code?! 😂

File: src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java [hunk 2]
```diff
@@ -182,11 +180,14 @@ private Cookie getCustomJwtCookieFromRequest(final HttpServletRequest request) {
...
+        final String saltId = stateCookieDto.getSaltId();
         final String encodedStateJson = oneLoginService.buildEncodedStateJson(
                 stateCookieDto.getRedirectUrl(),
-                stateCookieDto.getState()
+                stateCookieDto.getState(),
+                saltId
         );
-        final String hashedStateCookie = encryptionService.getSHA512SecurePassword(encodedStateJson);
+        final String hashedStateCookie = encryptionService.getSHA512SecurePassword(encodedStateJson, saltId);
+        encryptionService.deleteSalt(saltId);
```
Comment (chris-forbes) [inline src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java]: As discussed to wrap the delete in a check so we're not arbitrarily deleting salts

**Review by chris-forbes — CHANGES_REQUESTED** [review]

File: src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java (hunk 2 above)
Comment (ConorFayleTCO) [reply src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java]: done, thanks

File: src/main/java/gov/cabinetofice/gapuserservice/service/encryption/Sha512Service.java [hunk 3]
```diff
@@ -1,28 +1,58 @@
...
 
+import static gov.cabinetofice.gapuserservice.util.HelperUtils.generateSecureRandomString;
+import static gov.cabinetofice.gapuserservice.util.HelperUtils.generateUUID;
+
+@RequiredArgsConstructor
 @Service
 public class Sha512Service {
-    private String systemSalt;
+    private final SaltRepository saltRepository;
     private final int SALT_LENGTH = 255;
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/encryption/Sha512Service.java:21]: can you make this static?

`private final static int SALT_LENGTH = 255;`

File: src/main/java/gov/cabinetofice/gapuserservice/service/encryption/Sha512Service.java [hunk 4]
```diff
@@ -1,28 +1,58 @@
...
+        final Salt saltModel = Salt.builder().salt(salt).saltId(saltId).build();
+        this.saltRepository.save(saltModel);
+        return saltId;
+    }
+
+    public String generateAndStoreSalt() {
+        return storeSalt(generateSalt());
+    }
+
+    private String getSalt(String saltId) {
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/encryption/Sha512Service.java:38]: Can refactor this to use orElseThrow:

```
    private String getSalt(String saltId) {
        final Salt saltModel = this.saltRepository.findFirstBySaltIdOrderBySaltIdAsc(saltId)
                .orElseThrow(() -> new NonceExpiredException("Salt not found"));
        
        return saltModel.getSalt();
    }
```

File: src/main/java/gov/cabinetofice/gapuserservice/service/encryption/Sha512Service.java [hunk 5]
```diff
@@ -1,28 +1,58 @@
...
+        final Optional<Salt> saltModel = this.saltRepository.findFirstBySaltIdOrderBySaltIdAsc(saltId);
+        if (!saltModel.isPresent()) {
+            throw new NonceExpiredException("Salt not found");
+        }
+        return saltModel.get().getSalt();
+    }
+
+    public void deleteSalt(String saltId) {
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/encryption/Sha512Service.java:46]: You can call isPresent directly 

```
 public void deleteSalt(String saltId) {
        saltRepository.findFirstBySaltIdOrderBySaltIdAsc(saltId)
                .ifPresent(this.saltRepository::delete);
    }
```

Still functionally fine though!

File: src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java [hunk 6]
```diff
@@ -115,17 +114,25 @@ public RedirectView redirectAfterLogin(
...
         final StateCookieDto stateCookieDto = oneLoginService.decodeStateCookie(stateCookie);
         final String redirectUrl = stateCookieDto.getRedirectUrl();
 
         verifyStateAndNonce(decodedIdToken.getNonce(), stateCookieDto, state);
 
         final OneLoginUserInfoDto userInfo = oneLoginService.getOneLoginUserInfoDto(authToken);
-        oneLoginService.validateUserSub(decodedIdToken.getSub(), userInfo.getSub());
+
+        if (!Objects.equals(this.configProperties.getProfile(), "LOCAL")) {
+            // FIXME fix mocks so we don't have to ignore validation locally
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java:133]: Should remove the comment as it's an open source repo, could create a ticket for it instead?

File: src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java [hunk 7]
```diff
@@ -203,31 +215,31 @@ private void verifyStateAndNonce(final String nonce, final StateCookieDto stateC
...
                     keyValue("nonceFromDB", storedNonce.getNonceString()),
-                    keyValue("nonceCreatedAt", storedNonce.getCreatedAt()),
-                    keyValue("now", new Date()),
                     keyValue("stateFromResponse", state),
                     keyValue("hashedStateFromCookie", hashedStateCookie),
                     keyValue("stateFromCookie", encodedStateJson)
             );
-            throw new UnauthorizedClientException("User authorization failed, please try again");
-        } else if (!isStateVerified || !isNonceVerified) {
+            // TODO take action against malicious activity e.g. temp block user and send email
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java:229]: Can we create a ticket to capture this instead of the comment?

**Review by dylanwrightCO — APPROVED** [review]

**Review by dylanwrightCO — APPROVED** [review]

**Pull request merged**